src/blockdata/witness.rs

```diff
@@ -211,6 +213,17 @@ impl Witness {
         self.content[end_varint..].copy_from_slice(new_element);
     }
 
+    /// Pushes a DER-encoded ECDSA signature with a signature hash type as a new element on the
+    /// witness, requires an allocation.
+    pub fn push_bitcoin_signature(&mut self, signature: &ecdsa::SerializedSignature, hash_type: EcdsaSighashType) {
+        // Note that a maximal length ECDSA signature is 72 bytes, plus the sighash type makes 73
+        let mut sig = [0; 73];
+        sig[..signature.len()].copy_from_slice(&signature);
+        sig[signature.len()] = hash_type as u8;
+        self.push(&sig[..signature.len() + 1]);
+    }
```

Comment:
Do we want to protect against a malicious caller? `signature.len() returns a value that could have previously been set with `signature.set_len()`. A malicious call could cause this function to panic. Not sure if that matters?

One solution might be:
```
    pub fn push_bitcoin_signature(&mut self, signature: &ecdsa::SerializedSignature, hash_type: EcdsaSighashType) {
        // Note that a maximal length ECDSA signature is 72 bytes, plus the sighash type makes 73
        let mut sig = [0; 73];
        let len = signature.len();

        if len > 72 {
            // This can happen if a malicious caller has previously called
            // signature.set_len() with a length intentionally too big.
            return;
        }

        sig[..len].copy_from_slice(&signature);
        sig[len] = hash_type as u8;
        self.push(&sig[..=len]);
    }
```

Reply:
Perhaps we should just add a check to `set_len` instead?

Reply:
I'm not sure what is the security model you are thinking of.  If the attacker controls any code in the current process, aren't all bets off?  they can just read/write random memory locations, including keys and such.

Reply:
It also looks to me like `SerializedSignature` cannot be coerced into having a longer length. If it were, I'd advocate for a comment describing the panic, but I'm ok with it existing.

Reply:
> I'm not sure what is the security model you are thinking of

Sorry, I misspoke I should have said 'defensive programming' instead of 'malicious caller', my bad.

> I don't see a set_len() anywhere

Its `pub (crate)` so doesn't show up in those docs

> It also looks to me like SerializedSignature cannot be coerced into having a longer length.

Thanks for taking a look.